src/gnu/jftpgw/jftpgw.1
.\" Copyright (C) 2001 Joachim Wieland <joe@mcknight.de>
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place - Suite 330, Boston, MA 02111, USA.
.\"
.TH JFTPGW 1 "18 November 2001" "jftpgw" "User Manuals"
.SH NAME
jftpgw \- an FTP application level proxy/gateway
.SH SYNOPSIS
.B jftpgw
[OPTION]...
.SH DESCRIPTION
.PP
This manual page documents the program
.B jftpgw
which is an FTP proxy/gateway that uses the FTP protocol (unlike those FTP
proxies that fetch via FTP but provide the file via HTTP). You can use it to
make servers behind a firewall/NAT server (masquerading server) accessible
or to allow users behind such solutions to transfer files to and from the
outside of the LAN.
.SH OPTIONS
.TP
.I "\-h, \-\-help"
Display a short help text.
.TP
.I "\-e, \-\-encrypt"
Start the small, built-in password creation. jftpgw will read your input
password, output a crypt()ed version of this password and exit.
.TP
.I "\-f, \-\-configfile FILE"
Read FILE instead of the default configfile (instead of the built-in
location)
.TP
.I "\-s, \-\-single"
Run only a single thread of jftpgw (you won't need this option in the normal
use. It's nice for debugging, to trace the program or execute it within a
debugger).
.TP
.I "\-V, \-v, \-\-version"
Display the version number and exit.
.SH NOTES
.PP
Please note that the 0.13.x version is the first one that has a different
configuration than the versions released previously.  The new configuration
has a completely different (more flexible) configuration system and the
internal structure changed a lot.
.SH AVAILABILITY
.PP
The latest release is available on the jftpgw homepage.
.PP
Primary site: http://www.mcknight.de/jftpgw/
.SH "SEE ALSO"
.PP
Further information can be found in the
.I jftpgw.conf
file. The config file contains a description and some examples for every
option.
.TP
.I http://www.mcknight.de/jftpgw/
contains further information on how to configure the program for special
purposes and a list of all configuration options.
.SH AUTHORS
.PP
Joachim Wieland <joe@mcknight.de> is the original author and current
maintainer of jftpgw.



CHANGES
		CHANGES
		Last Modified: $Date$ (GMT)

2004/05/12

  add gentraffic model.
	This generate network traffic by ttcp.

2004/02/09

  conf/1.6/kernel.i386
	clean up logic.
	sync GENERIC file w/ 1.6.1.

2004/02/07

  enable IPv6 multicast model by default.

  remove model/[a-z]* (old directories)
	now we use model/$version/$model

2004/02/06

  new feature: $bridge_flap

	if $bridge_flap = yes, fdgw simulates line trouble by running
	the folloging command repeatedly :-).

	   while true
	   do
		ifconfig bridge0 down; sleep 300;
		ifconfig bridge0 up;   sleep 300;
  	   done

2004/02/05

  support bridge model. 

  clean up conf/1.6/kernel.i386
	re-create kernel config file from 1.6 source.
	add BRIDGE for bridge model.

2004/01/28

  update pim6sd
	sync w/ pkgsrc pim6sd-20031021a

  compilable on NetBSD 1.6.1/i386

	fix kernel options in conf/1.6/FDGW{,6}.
		disable ipip if 1.6

	model/1.6/
		use modified syslogd source which is prepared
		on the fly in building stage.

	directory hierarchy changed to include version dependence.
	 	model/${model} -> model/${version}/${model}
		create model/1.6 and conf/1.6
			imported from model/1.5 and conf/1.5 respectively.

	fix lndir targets if on 1.6.

2003/07/02

   conf/1.5/etc/rc.subr
	bug fix: reverse return value of check_prog_exist()

2003/01/17

   add vpn-racoon model
	IPSec VPN by IKE.
	we use racoon as IKE daemon.

   clean up the usage of files and directories.
	not use src/gnu/ as a work dir in compilation
	use src/pkg/ for external packages.
	renamed: src/distinfo -> src/gnu/distinfo
	create src/patches/
		This dir holds patches for packages, 
		imported from pkgsrc mainly.

2003/01/10

   new rc.conf generator
	clean up model/*/rc.conf.
	add src/merge_conf.awk 

   clean up rc.conf(s) in conf/ and model/
	implement new merging process by awk.
	it creates /etc/rc.conf on image by
	both model/*/rc.conf and defaults/rc.conf.

   src/Makefile 
	import model/*/rc.conf as obj.$arch.$model/rc.conf.
	pass _CONF_DIR to sub make process (but not needed ?).

   src/Makefile.bootfloppy 
	new merge process by awk which merges obj.$arch.$model/rc.conf
	and defaults/rc.conf.

	import local/model/$model/ files if exists.

2003/01/09

   clean up conf/1.5/etc/
	install etc/defaults/rc.conf as /etc/rc.conf on image.

   clean up, fix comments

2002/12/09

   /.profile
	PATH=/sbin:/bin:/usr/bin:/usr/sbin:/:/usr/pkg/sbin:/usr/pkg/bin

2002/08/05

   conf/1.5/etc/
	add rtsold, rtadvd (disabled by default)

   add IPv6 multicast model (mcastrouter6)

   modify riprouter model
	add rtadvd, rtsold. 
	nuke mount_msdos, dhcpd. 
	nuke -lipsec

   import pim6sd source

2002/07/31

   enable BIOSBOOT options
	make BIOSBOOT=biosboot_com0.sym ...

2002/06/20

   change the directory hierarchy to handle version dependence
	conf/ -> conf/1.5
	src.$arch.$model -> obj.$arch.$model

2002/04/25

   add variables to control IP forwarding
	$ip4forwarding
	$ip6forwarding

   clean up
	prepare rc.conf for each model
	clean up riprouter model

2002/04/24

   add riprouter model to handle routing by RIP in both IPv4 and IPv6 cases
	primary target is a IPv6 router, which needs RIPv2 only now.

2002/04/22

   apply patch-ae from pkgsrc/sysutils/fdgw (requested by wennmach@nebsd.org)
	fix patch-ae bug (we should lndir /sys/conf too)

2002/04/18

   add BGP router model based on zebra

   verify checksum

2002/04/17

   updated squid from 2.4 stable 4 to stable 6.

   use our lib/libtransparent/ipfilter.c not jftpgw builtin one

   include sysctl into proxybox model

2002/04/16

   change /etc directory
	copy /etc on /dev/fd0a to /etc/ on md0a and use it as /etc not
	/conf/etc

2002/02/25

   install pwd.db after booted to decrease /etc size on fd0a

2002/02/23

   show statistics (Makefile src/stat.sh)
	results by df -i for ramdisk and floppy

   remove /etc/spwd.db but hold /etc/pwd.db

2002/02/22

   minor bug fix for proxymodel
	update squid to 2.4 stable 4.
	tune up squid parameters
	use /usr/pkg/etc/squid/ for squid configuration files.
	install config files to /usr/pkg/etc/squid/ from /etc/squid/
	in boot time.

2002/02/21

   minor bug fixes proxymodel around squid
	log file rotations to avoid file system full
	install icons/
	fix permissions et.al.

   get tarball and extract it.
	we should not hold sources in fdgw itself to shlink fdgw source tree.
	o squid
	o jftpgw

2002/02/20

   o.k. proxybox works !
	o add stone as functional proxies
	o define ifconfig.lo0
	o fix squid.conf fules

   rc.router
	define the following proxies:
	o smtp_proxy
	o https_proxy
	o anoncvs_proxy
	o irc_proxy

   add stone for functional tcp proxy
	+ libwrap
	+ ip filter support
	+ rename $0 by setproctitle()

2002/02/19

   improve proxybox model templates

   src/etc/squid
	master.sh: fix path, squid + "-s" option
	bug fix around acl local src

2002/02/18

   clarify model dependence / indepentdence
	o install model specific rc.conf from model/$model/
	o model specific rc.conf precedes default one

   prepare proxy initialization codes in src/etc/$proxy/
	for example, src/etc/squid/master.sh

2002/02/15

   model/proxybox/
	fix path:
	o /usr/pkg/sbin/squid (not /bin/squid)
	o /usr/pkg/sbin/jftpgw not /bin/
	add bin/sleep
	install unlinkd
	install mime.conf
	create /etc/squid/icons/
	remove /usr/pkg/bin/tproxy
	squid is modified to use ipf transparency

   change configure options for squid
   do not build transproxy, which is not needed.

   conf/etc/rc.router
	add definitions for http_proxy ftp_proxy ssh_proxy pop_proxy

2002/02/13

   add ipsec configuration

2002/02/12

   bug fix ramdiskbin.conf
	remove -lbz2, which is not used

2002/02/11

   bug fix paths for configuration files
	installed into /etc on fd0a not vnd0a

   shrink squid.conf for floppy size limitation

   rename proxyfw -> proxybox model.
   clean up

2002/02/10

   import stone, jftpgw, transproxy

2002/02/09

   add proxyfw model (prototype)
	enable squid
	import squid source in src/gnu/squid/


src/gnu/jftpgw/jftpgw.c
/* 
 * Copyright (C) 1999-2001 Joachim Wieland <joe@mcknight.de>
 * 
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place - Suite 330, Boston, MA 02111, USA.
 */

#include <ctype.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/resource.h>
#include <unistd.h>
#ifdef HAVE_GETOPT_LONG
#include <getopt.h>
#else
#include "support/getopt.h"
#include "support/getopt.c"
#include "support/getopt1.c"
#endif

#include "jftpgw.h"


void reset_loginfo(struct loginfo_st*);
void print_version(void);
void print_help(void);

struct clientinfo clntinfo;
struct serverinfo srvinfo;
extern struct loginfo_st loginfo;
extern struct log_cmd_st lcs;
extern struct uidstruct runasuser;
extern struct connliststruct* connected_clients;
extern struct slist_t* passcmd_white_list, *passcmd_black_list;
extern int should_read_config;
int timeout;


static struct option const long_option_arr[] =
{
	{ "single",     no_argument,       NULL, 's' },
	{ "inetd",      no_argument,       NULL, 'i' },
#ifdef HAVE_LIBWRAP
	{ "tcpwrap",    no_argument,       NULL, 't' },
#endif
	{ "encrypt",    no_argument,       NULL, 'e' },
	{ "version",    no_argument,       NULL, 'v' },
	{ "version",    no_argument,       NULL, 'V' },
	{ "help",       no_argument,       NULL, 'h' },
	{ "configfile", required_argument, NULL, 'f' },
	{ NULL,         0,                 NULL,  0  }
};

int main(int argc, char** argv) {

	/* We bind to 0.0.0.0, but our real IP is the one returned by the
	 * PASV command */

	struct sigaction sa, cf;
	int ret;

#ifdef RLIMIT_CORE
	/* set RLIMIT_CORE to 0 */
	struct rlimit rl = { 0, 0 };

	if (setrlimit(RLIMIT_CORE, &rl) < 0) {
		perror("setrlimit");
		return -1;
	}
#endif
	/* set the name of the config file to an initial value, it may be
	 * overwritten later
	 * */
	srvinfo.conffilename = strdup(DEFAULTCONFFILE);
	enough_mem(srvinfo.conffilename);

	/* No login done yet */
	clntinfo.login.stage = LOGIN_ST_NOT_CONNECTED;

	/* default: Multithread, i.e. fork for each connection */
	srvinfo.multithread = 1;
	srvinfo.tcp_wrapper = 0;
	srvinfo.servertype = SERVERTYPE_STANDALONE;
	srvinfo.main_server_pid = 0;
	srvinfo.chrooted = 0;

	if (strrchr(argv[0], '/')) {
		srvinfo.binaryname = strdup(strrchr(argv[0], '/') + 1);
	} else {
		srvinfo.binaryname = strdup(argv[0]);
	}
	enough_mem(srvinfo.binaryname);

	/* parse the command line */
	while ((ret=getopt_long(argc, argv, "isetvVhf:", long_option_arr, NULL)) != EOF) {
		switch (ret) {
			case 0: break;
			case 'i':
				srvinfo.servertype = SERVERTYPE_INETD;
				srvinfo.multithread = 0;
				break;
			case 's':
				srvinfo.multithread = 0;
				break;
#ifdef HAVE_LIBWRAP
			case 't':
				srvinfo.tcp_wrapper = 1;
				break;
#endif
			case 'e':
				encrypt_password();
				exit(0);
			case 'v':
			case 'V':
				print_version();
				exit(0);
			case 'h':
				print_help();
				exit(0);
			case 'f':
				if (set_conffilename(optarg) < 0) {
					log(2, "something is wrong with the name of the configuration file");
					return 1;
				}
				break;
			default:
				break;
		}
	}

	/* Read the configuration */
	memset(&loginfo, 0, sizeof(struct loginfo_st));
	loginfo.debuglevel = 6;
	srvinfo.ready_to_serve = SVR_LAUNCH_CMDLINE;
	srvinfo.chrootdir_saved = (char*) 0;

	ret = read_config(srvinfo.conffilename);
	if (ret) {
		return 1;
	}

	/* Drop privileges right after the start of the program. Right after
	 * reading the configuration file */

	if (stage_action("start") < 0) {
		return -1;
	}

	/* init the logfiles */
	if (changeid(PRIV, UID, "log_init()")    < 0) { return -1; }
	if (log_init()                           < 0) {
		changeid(UNPRIV, EUID, "log_init()");
		return -1;
	}
	if (changeid(UNPRIV, EUID, "log_init()") < 0) { return -1; }

	srvinfo.ready_to_serve = SVR_LAUNCH_LOGFILES;

	/* initiate a few values */
	clntinfo.transparent = TRANSPARENT_NO;
	clntinfo.destinationport = 0;

	/* Install the signal handlers */

	/* for SIGCHLD install just the reap function
	 *
	 * the register/unregister thing is installed after we've bound
	 * successfully */

	sa.sa_handler = reap_chld_info;
	sigemptyset (&sa.sa_mask);
#ifndef WINDOWS
	sa.sa_flags = SA_RESTART;
#endif
	sigaction (SIGCHLD, &sa, 0);

	cf.sa_handler = read_default_conf;
	should_read_config = 0;
	sigemptyset(&cf.sa_mask);
#ifndef WINDOWS
	cf.sa_flags = SA_RESTART;
#endif
	sigaction (SIGHUP, &cf, 0);

	cf.sa_handler = terminate;
	sigemptyset(&cf.sa_mask);
	sigaction (SIGTERM, &cf, 0);
	sigaction (SIGQUIT, &cf, 0);
	sigaction (SIGABRT, &cf, 0);

	/* Ignore SIGALRM */
	sa.sa_handler = SIG_IGN;
	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;
	sigaction(SIGALRM, &sa, 0);

	clntinfo.user = clntinfo.pass = clntinfo.destination = (char*) 0;
	clntinfo.before_forward.user = clntinfo.before_forward.destination
								= (char*) 0;
	clntinfo.anon_user = (char*) 0;
	clntinfo.throughput = 0;
	clntinfo.boundsocket_list = (int*) 0;
	clntinfo.server_ip = clntinfo.client_ip = clntinfo.addr_to_server
		= clntinfo.addr_to_client = (unsigned long int) UINT_MAX;

	atexit(closedescriptors);

	if (srvinfo.servertype == SERVERTYPE_STANDALONE &&
		(ret = waitclient(config_get_option("listen"),
							&clntinfo)) < 0) {
		/* An error occured */
		return ret;
	}
	if (srvinfo.servertype == SERVERTYPE_INETD) {
		close(1);
		dup2(0, 10);
		close(0);
		if (inetd_connected(10, &clntinfo) < 0) {
			return -1;
		}
	}
	ret = handle_login(&clntinfo);
	log(8, "Exited from handle_login, ret: %d", ret);
	if (!ret) {
		ret = handle_cmds(&clntinfo);
		log(8, "Exited from handle_cmds");
	}

	log(8, "Exiting");
	if (ret) {
		return 1;
	} else {
		return 0;
	}
}


int daemonize()
{
	int childpid;

	if( (childpid = fork()) < 0) return(-1);
	else if(childpid > 0) exit(0);

	errno = 0;
	chdir("/");
	setsid();
	return(0);
}


void read_default_conf(int signo) {
	should_read_config = 1;
}

int reread_config() {
	int ret;
	sigset_t sigset, oldset;
	log(6, "SIGHUP received - Rereading config file");

	/* block other SIGHUP signals */

	sigemptyset(&sigset);
	sigemptyset(&oldset);
	sigaddset(&sigset, SIGCHLD);
	while ((ret = sigprocmask(SIG_BLOCK, &sigset,
		&oldset)) < 0 && errno == EINTR)  {}
	if (ret < 0) {
		log(2, "Error blocking signals: %s", strerror(errno));
	}
	config_delete_config();
	destroy_active_portrange();
	destroy_passive_portrange();
	ret = read_config(srvinfo.conffilename);
	if (ret) {
		/* the following line is a BADF if we had a SIGHUP */
		log(1, "Error rereading config file. Exiting.");
		exit(2);
	}
	/* re-register all connected clients */
	config_counter_add_connected(connected_clients);
	reset_loginfo(&loginfo);
	if (stage_action("reread") < 0) {
		return -1;
	}
	should_read_config = 0;
	while ((ret = sigprocmask(SIG_UNBLOCK, &sigset,
		&oldset)) < 0 && errno == EINTR) {}
	if (ret < 0) {
		log(2, "Error unblocking signal mask: %s", strerror(errno));
		return -1;
	}
	return 0;
}

void terminate (int signo) {
	/* exit is not POSIX-reentrant but in SVR4 SVID */
	exit(0);
}


int stage_action(const char* stage) {
	char* reason = char_append("stage_action() in stage ", stage);

	changeid(PRIV, UID, reason);
	free(reason);
	if (change_root(stage) < 0) {
		changeid(UNPRIV, EUID, "change_root() failed");
		return -1;
	}
	if (dropprivileges(stage) < 0) {
		changeid(UNPRIV, EUID, "dropprivileges() failed");
		return -1;
	}
	changeid(UNPRIV, EUID, "log_init()");
	return 0;
}


void print_version(void) {
	printf(PACKAGE" v"JFTPGW_VERSION);
#ifdef HAVE_CRYPT
	printf("  -  crypt support enabled");
#else
	printf("  -  without crypt support");
#endif


#ifdef HAVE_LINUX_NETFILTER_IPV4_H
	printf("  -  netfilter support enabled");
#else
	printf("  -  without netfilter support");
#endif
	printf("\n");

#ifdef HAVE_LIBWRAP
	printf("libwrap support enabled");
#else
	printf("without libwrap support");
#endif

#ifdef HAVE_SIOCGIFADDR
	printf(" - can get IPs from interfaces");
#else
	printf(" - can't get IPs from interfaces");
#endif

#ifdef HAVE_ICMP_SUPPORT
	printf(" - ICMP support");
#else
	printf(" - no ICMP support");
#endif
	printf("\n");


}

void print_help(void) {
	print_version();
	printf("usage: jftpgw [OPTION]\n\n");
	printf("Valid options:\n");
	printf("  -h, --help                Display this help text\n");
	printf("  -e, --encrypt             Use jftpgw to obtain an encrypted password\n");
	printf("  -f, --configfile file     Load file instead of default config file\n");
	printf("  -s, --single              Run jftpgw single threaded (do not fork)\n");
	printf("  -i, --inetd               Run jftpgw from inetd superserver\n");
#ifdef HAVE_LIBWRAP
	printf("  -t, --tcpwrap             Use libwrap for access control\n");
#endif
	printf("  -V, -v, --version         Display the version\n");
	printf("\nReport bugs to Joachim Wieland <joe@mcknight.de>\n");
}


void removepidfile(void) {
	const char* option;
	if (getpid() != srvinfo.main_server_pid) {
		/* the program has not become a daemon */
		return;
	}

	option = config_get_option("pidfile");
	if (option) {
		int i;
		if (changeid(PRIV, UID,
			"Changing ID to root (unlink pidfile)") < 0) {
			return;
		}
		i = unlink(option);
		if (i) {
			log(3, "Could not unlink the pidfile %s", option);
		}
		if (changeid(UNPRIV, EUID,
				"Changing id back (deleting pidfile)") < 0) {
			return;
		}
	}
}

void sayterminating(void) {
	log(6, "jftpgw terminating");
}

void closedescriptors(void) {
	int i;
	log(9, "In closedescriptors()");

	/* free the log info structure. Free the members, the structure for
	 * itself is on the stack */
	/* lcs.cmd must not be freed, it's   lcs.cmd = buffer;  */
	/* the same for lcs.filename */
	if (lcs.svrip)    { free(lcs.svrip);    }
	if (lcs.clntip)   { free(lcs.clntip);   }
	if (lcs.ifipclnt) { free(lcs.ifipclnt); }
	if (lcs.ifipsvr)  { free(lcs.ifipsvr);  }

	if (lcs.userlogin)      { free(lcs.userlogin);      }
	if (lcs.usereffective)  { free(lcs.usereffective);  }
	if (lcs.userforwarded)  { free(lcs.userforwarded);  }

	if (srvinfo.conffilename) { free(srvinfo.conffilename); }

	/* close the logfiles and delete the structures */
	reset_loginfo(&loginfo);
	free(srvinfo.chrootdir_saved);

	config_delete_config();
	config_delete_backup();

	/* maybe these variables are not yet allocated */
	if (clntinfo.user) {
		free(clntinfo.user);
	}
	if (clntinfo.pass) {
		free(clntinfo.pass);
	}
	if (clntinfo.destination) {
		free(clntinfo.destination);
	}
	if (runasuser.username) {
		free(runasuser.username);
		runasuser.username = 0;
	}
	if (runasuser.groupname) {
		free(runasuser.groupname);
		runasuser.groupname = 0;
	}
	if (clntinfo.anon_user) {
		free(clntinfo.anon_user);
		clntinfo.anon_user = (char*) 0;
	}
	if (clntinfo.before_forward.user) {
		free(clntinfo.before_forward.user);
		clntinfo.before_forward.user = (char*) 0;
	}
	if (clntinfo.before_forward.destination) {
		free(clntinfo.before_forward.destination);
		clntinfo.before_forward.destination = (char*) 0;
	}
	free_errstr();

	free(srvinfo.binaryname);

	slist_destroy(passcmd_white_list);
	slist_destroy(passcmd_black_list);

	destroy_active_portrange();
	destroy_passive_portrange();

	if (clntinfo.boundsocket_list) {
		for (i = 0; i < clntinfo.boundsocket_niface; i++) {
			close(clntinfo.boundsocket_list[i]);
		}
		free(clntinfo.boundsocket_list);
	}
	close(clntinfo.clientsocket);
	close(clntinfo.serversocket);
}


char* chrooted_path(const char* path) {
	char* newpath;
	const char* p, *start;
	const char* chrootdir = config_get_option("changerootdir");

	if (!chrootdir) {
		/* Try the saved path */
		chrootdir = srvinfo.chrootdir_saved;
	} else {
		if (!srvinfo.chrootdir_saved) {
			/* save it */
			srvinfo.chrootdir_saved = strdup(chrootdir);
			enough_mem(srvinfo.chrootdir_saved);
		}
	}
	/* okay, maybe we are chrooted, see if the logfile path lies in the
	 * chrooted dir */
	if (path && chrootdir && srvinfo.chrooted) {
		p = strstr(path, chrootdir);
		if (p == path) {
			/* Yeah, we're inside the chrooted dir, strip it
			 * off.
			 * Calculate where to split, make sure, the new path
			 * starts with a slash. */
			start = path + strlen(chrootdir);
			while (start > path && *start == '/') {
				start --;
			}
			start++;
			newpath = strdup(start);
			enough_mem(newpath);
			/* just for the beauty... */
			char_squeeze(newpath, '/');
			return newpath;
		}
	}
	/* maybe we are not chrooted or the chrootdir is not part of the
	 * path...*/
	newpath = strdup(path);
	enough_mem(newpath);
	/* just for the beauty... */
	char_squeeze(newpath, '/');

	return newpath;
}

